### PR TITLE
*: remove mysql.TypeDecimal

### PR DESCRIPTION
**Pull request by wangggong**
Signed-off-by: Wang Ruichao <wangruichao2014@xiaochuankeji.cn>

### What problem does this PR solve?

close https://github.com/pingcap/parser/issues/767
cc https://github.com/pingcap/tidb/issues/15328

Problem Summary:

### What is changed and how it works?

What's Changed:

Remove mysql.TypeDecimal, instead mysql.TypeNewDecimal / mysql.TypeUnspecified is used.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

Nope. Please tell me if there's any test to add or code to modify. Thanks.

### CODE REVIEW

**Comment by wangggong** [comment]
Note that in my opinion some `mysql.TypeDecimal` can be replaced into `mysql.TypeNewDecimal` instead of `mysql.TypeUnspecified`, while some test with the real TiKV cannot pass. To make CI happy, I commit another commit. I'm curious whether the 1st commit is reasonable. If so, I'll try to debug & fix the test with TiKV.

File: cmd/importer/db.go [hunk 1]
```diff
@@ -277,7 +277,7 @@ func genColumnData(table *table, column *column) (string, error) {
 
 		data = append(data, '\'')
 		return string(data), nil
-	case mysql.TypeNewDecimal, mysql.TypeDecimal:
```
Comment (XuHuaiyu) [inline cmd/importer/db.go:280]: We should remove the TypeDecimal here, it's misuse

File: ddl/index.go [hunk 2]
```diff
@@ -202,7 +202,7 @@ func getIndexColumnLength(col *model.ColumnInfo, colLen int) (int, error) {
 			return mysql.DefaultLengthOfMysqlTypes[mysql.TypeFloat], nil
 		}
 		return mysql.DefaultLengthOfMysqlTypes[mysql.TypeDouble], nil
-	case mysql.TypeDecimal, mysql.TypeNewDecimal:
```
Comment (XuHuaiyu) [inline ddl/index.go:205]: We should remove the TypeDecimal here, it's misuse

File: expression/builtin_time.go [hunk 3]
```diff
@@ -4867,7 +4867,7 @@ func (c *timestampFunctionClass) getFunction(ctx sessionctx.Context, args []Expr
 	}
 	isFloat := false
 	switch args[0].GetType().Tp {
-	case mysql.TypeFloat, mysql.TypeDouble, mysql.TypeDecimal:
```
Comment (XuHuaiyu) [inline expression/builtin_time.go:4870]: We need to change this TypeDecimal to TypeNewDecimal

File: expression/generator/other_vec.go [hunk 4]
```diff
@@ -379,8 +379,8 @@ var vecBuiltin{{ .Category }}GeneratedCases = map[string][]vecExprBenchCase {
 					{Value: types.NewFloat64Datum(0.2), RetType: types.NewFieldType(mysql.TypeFloat)},
 				{{- end }}
 				{{- if eq .Input.ETName "Decimal" }}
-					{Value: types.NewDecimalDatum(types.NewDecFromInt(10)), RetType: types.NewFieldType(mysql.TypeDecimal)},
```
Comment (XuHuaiyu) [inline expression/generator/other_vec.go:382]: This should be changed to TypeNewDecimal

File: expression/generator/other_vec.go [hunk 5]
```diff
@@ -379,8 +379,8 @@ var vecBuiltin{{ .Category }}GeneratedCases = map[string][]vecExprBenchCase {
...
-					{Value: types.NewDecimalDatum(types.NewDecFromInt(20)), RetType: types.NewFieldType(mysql.TypeDecimal)},
```
Comment (XuHuaiyu) [inline expression/generator/other_vec.go:383]: This should be changed to TypeNewDecimal

File: types/etc.go [hunk 6]
```diff
@@ -68,7 +68,7 @@ func IsTypeTime(tp byte) bool {
 func IsTypeNumeric(tp byte) bool {
 	switch tp {
 	case mysql.TypeBit, mysql.TypeTiny, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeNewDecimal,
-		mysql.TypeDecimal, mysql.TypeFloat, mysql.TypeDouble, mysql.TypeShort:
```
Comment (XuHuaiyu) [inline types/etc.go:71]: We should change this to TypeNewDecimal

File: types/etc_test.go [hunk 7]
```diff
@@ -323,7 +323,7 @@ func (s *testTypeEtcSuite) TestIsTypeNumeric(c *C) {
 	res = IsTypeNumeric(mysql.TypeNewDecimal)
 	c.Assert(res, Equals, true)
 
-	res = IsTypeNumeric(mysql.TypeDecimal)
```
Comment (XuHuaiyu) [inline types/etc_test.go:326]: We should change this to TypeNewDecimal

File: types/field_type.go [hunk 8]
```diff
@@ -763,7 +763,7 @@ var fieldTypeMergeRules = [fieldTypeNum][fieldTypeNum]byte{
 	/* mysql.TypeYear -> */
 	{
 		//mysql.TypeDecimal      mysql.TypeTiny
```
Comment (XuHuaiyu) [inline types/field_type.go:765]: This comment also need to be changed

File: expression/builtin_cast_test.go [hunk 9]
```diff
@@ -253,7 +253,7 @@ func (s *testEvaluatorSuite) TestCastXXX(c *C) {
 
 	// cast(bad_string as decimal)
 	for _, s := range []string{"hello", ""} {
-		f = BuildCastFunction(ctx, &Constant{Value: types.NewDatum(s), RetType: types.NewFieldType(mysql.TypeDecimal)}, tp)
```
Comment (XuHuaiyu) [inline expression/builtin_cast_test.go:256]: This needs to be changed to TypeNewDecimal

File: cmd/importer/db.go (hunk 1 above)
Comment (wangggong) [reply cmd/importer/db.go:280]: revert done, thx~

ditto.

File: expression/expr_to_pb_test.go [hunk 10]
```diff
@@ -147,7 +147,6 @@ func (s *testEvaluatorSuite) TestColumn2Pb(c *C) {
 	colExprs = append(colExprs, dg.genColumn(mysql.TypeEnum, 3))
 	colExprs = append(colExprs, dg.genColumn(mysql.TypeGeometry, 4))
 	colExprs = append(colExprs, dg.genColumn(mysql.TypeUnspecified, 5))
-	colExprs = append(colExprs, dg.genColumn(mysql.TypeDecimal, 6))
```
Comment (XuHuaiyu) [inline expression/expr_to_pb_test.go:150]: Why remove this?

Comment (wangggong) [reply expression/expr_to_pb_test.go:150]: Since if `mysql.TypeDecimal` is replaced into `mysql.TypeUnspecified`, the test will duplicate.

Comment (XuHuaiyu) [reply expression/expr_to_pb_test.go:150]: Maybe we can change this TypeDecimal to TypeNewDecimal?

Comment (wangggong) [reply expression/expr_to_pb_test.go:150]: Emmmm... The test for `mysql.TypeNewDecimal` has been added in the following code. Add it will also duplicate.

**Comment by tangenta** [comment]
@wangggong We should update the parser dependency through
```
go get -v github.com/pingcap/parser@1259bef
go mod tidy
```

**Comment by wangggong** [comment]
> @wangggong We should update the parser dependency through
> 
> ```
> go get -v github.com/pingcap/parser@1259bef
> go mod tidy
> ```

done, thx~ However there're some structs deleted in `parser` repo and still be used in `tidb` repo, making compile cannot be completed. I will try to fix it.

**Comment by wangggong** [comment]
> @wangggong We should update the parser dependency through
> 
> ```
> go get -v github.com/pingcap/parser@1259bef
> go mod tidy
> ```

done. @tangenta @XuHuaiyu @AilinKid PTAL

**Comment by AilinKid** [comment]
there is some max display length problem in TiDB-Test repo, I've added a parser variable to control the behavior. Focusing on https://github.com/pingcap/parser/pull/949, update your parser again when it merged. Thanks

**Comment by AilinKid** [comment]
go get -v github.com/pingcap/parser@5fe0b0fdf629026232c91b80ea50e734315bf9e3
go mod tidy

**Comment by wangggong** [comment]
> there is some max display length problem in TiDB-Test repo, I've added a parser variable to control the behavior. Focusing on [pingcap/parser#949](https://github.com/pingcap/parser/pull/949), update your parser again when it merged. Thanks

OK, done~

**Comment by AilinKid** [comment]
The CI failure seems nothing to do with this PR.
This PR has blocked some pr with newer parser ( make dev will fail), that's why we merge it fristly here.

**Pull request merged**